lib/elixir/lib/gen_server/loop.ex

```diff
@@ -0,0 +1,392 @@
+        init_ok(starter, parent, dbg, name, mod, state, await)
+      {:stop, reason} ->
+        init_stop(starter, reason, gen_name)
+      :ignore ->
+        init_stop(starter, :normal, gen_name, :ignore)
+      other ->
+        init_stop(starter, {:bad_return_value, other}, gen_name)
+    end
+  end
+
+  ## :proc_lib callbacks
+
+  @spec continue(pid(), [:sys.dbg_opt], term(), module(), term()) :: no_return()
+  def continue(parent, dbg, name, mod, state) do
+    # must be message in queue as woke up from hibernation
+    receive do
+      msg ->
+        handle(msg, parent, dbg, name, mod, state, :hibernate)
+    end
+  end
+
+  ## :sys callbacks
+
+  def system_continue(parent, dbg, {name, mod, state, await}) do
+    loop(parent, dbg, name, mod, state, await)
+  end
+
+  def system_get_state({_, _, state, _}) do
+    {:ok, state}
+  end
+
+  def system_replace_state(replace, {name, mod, state, await}) do
+    state = replace.(state)
+    {:ok, state, {name, mod, state, await}}
+  end
+
+  def system_code_change({name, mod, state, await}, _, oldvsn, extra) do
+    try do
+      apply(mod, :code_change, [oldvsn, state, extra])
+    catch
+      :throw, value ->
+        :erlang.raise(:error, {:nocatch, value}, System.stacktrace())
+    else
+      {:ok, state} ->
+        {:ok, {name, mod, state, await}}
+    end
+  end
+
+  def format_status(opt, [pdict, status, parent, dbg, {name, mod, state, _}]) do
+    header = 'Status for generic server: #{inspect(name)}'
+    events = :sys.get_debug(:log, dbg, [])
+    data = [{'Status', status}, {'Parent', parent}, {'Logged events', events}]
+    [header: header, data: data] ++ format_status(mod, opt, pdict, state)
+  end
+
+  def system_terminate(reason, _, dbg, {name, mod, state, _}) do
+    try do
+      throw(:terminate)
+    catch
+      :throw, :terminate ->
+        stack = System.stacktrace()
+        handle_exception(:exit, reason, stack, nil, dbg, name, mod, state)
+    end
+  end
+
+  ## Helpers
+
+  defp server_parent(:self),  do: self()
+  defp server_parent(parent), do: parent
+
+  defp server_name(name) when is_pid(name), do: name
+  defp server_name({:local, name}),         do: name
+  defp server_name({:global, name}),        do: name
+  defp server_name({:via, _, name}),        do: name
+
+  defp server_debug(name, opts) do
+    dbg_opts = Keyword.get(opts, :debug, [])
+    try do
+      :sys.debug_options(opts)
+    catch
+      _, _ ->
+        format = '** Generic server ~p started with invalid debug options: ~p~n'
+        :error_logger.format(format, [name, dbg_opts])
+        []
+    end
+  end
+
+  defp debug([], _, _, _, _), do: []
+  defp debug(dbg, event, name, mod, state) do
+    :sys.handle_debug(dbg, &debug_inspect/3, {name, mod, state}, event)
+  end
+
+  defp debug_inspect(device, event, {name, mod, state}) do
+    msg = ["** (DEBUG from ", inspect(name), ?), ?\s, inspect(mod),
+           " with state ", inspect(state) | debug_format(event)]
```

Comment:
I'm not sure this format will be really helpful for the user. The server's inner state can be a really large term, and printing it inline, might make it hard to see what is happening. I'm worried it will only clutter the debug messages making it harder to see what is happening instead of helping.
At least I think the state should go pretty-printed at the very end of the debug message.